src/Database/SqlserverCompiler.php

```diff
@@ -49,10 +49,30 @@ class SqlserverCompiler extends QueryCompiler
      * @inheritDoc
      */
     protected $_selectParts = [
-        'select', 'from', 'join', 'where', 'group', 'having', 'order', 'offset',
+        'with', 'select', 'from', 'join', 'where', 'group', 'having', 'order', 'offset',
         'limit', 'union', 'epilog',
     ];
 
+    /**
+     * Helper function used to build the string representation of a `WITH` clause,
+     * it constructs the CTE definitions list without generating the `RECURSIVE`
+     * keyword that is neither required nor valid.
+     *
+     * @param \Cake\Database\Expression\CommonTableExpression[] $parts List of CTEs to be transformed to string
+     * @param \Cake\Database\Query $query The query that is being compiled
+     * @param \Cake\Database\ValueBinder $generator The placeholder generator to be used in expressions
+     * @return string
+     */
+    protected function _buildWithPart(array $parts, Query $query, ValueBinder $generator): string
+    {
+        $expressions = [];
+        foreach ($parts as $expression) {
+            $expressions[] = $expression->sql($generator);
+        }
+
+        return sprintf('WITH %s ', implode(', ', $expressions));
```

Comment:
We don't need this override. Users shouldn't create a recursive cte with sqlserver. It won't execute properly anyway.

Reply:
SQL Server does support recursive CTEs, it's just that unlike other DBMS (and contrary to the SQL standard) it just doesn't use the `RECURSIVE` keyword.

If we'd drop this, then we're forcing people to write rather DBMS specific queries when it's not really necessary.

Reply:
I was not familiar with sql server recursive notation. As long as we aren't hiding bad sql then this looks good to me.

I think another approach is modifying the expression on compilation ... but this looks clean.